File: drivers/term.c
#include "term.h"

static struct cursor cur = {0, 0, (uint16_t *)VGA_MEM_START};

void term_cursor_move(uint16_t steps)
{
	cur.curp += steps;
	cur.x += steps;
	
	// wrap to next line if line width increases 80 
	if (cur.x / VGA_X > 0) {
		cur.x = cur.x % VGA_X;
		cur.y += cur.x / VGA_X;
	}
	return;
}

void term_cursor_set(uint8_t x, uint8_t y)
{
	cur.x = x;
	cur.y = y;
	cur.curp = ((uint16_t*)VGA_MEM_START) + VGA_OFFSET(x, y);
	return;
}

void term_clear()
{
	vga_cls();
	term_cursor_set(0, 0);
}

void term_print(char* str)
{
	char* ch = str;
	while (*ch) {
		switch (*ch) {
			case '\n':
				// set the cursor to beginning of next line
				term_cursor_set(0, cur.y + 1);
				break;
			case '\r':
				// set the cursor to beginning of current line
				term_cursor_set(0, cur.y);
				break;
			default:
				// not an escape character
				vga_put(*ch, cur.curp);
				term_cursor_move(1);
				break;
		}	
		ch++;
	}

	return;
}

File: src/drivers/term.rs
use crate::drivers::vga;
use lazy_static::lazy_static;
use spin::Mutex;
use core::fmt;

pub struct TermCursor {
   row: usize,
   col: usize,
}

impl TermCursor {

	fn set_cursor(&mut self, row: usize, col: usize)
	{
		self.row = row;
		self.col = col;
	}


	fn move_cursor(&mut self, steps: usize) -> () {
		self.col += steps;
		
		// wrap to next line if line width increases 80 
		if (self.col / vga::VGA_COLS > 0) {
			self.row += self.col / vga::VGA_COLS;
			self.col = self.col % vga::VGA_COLS;
		}
	}

	pub fn cls(&mut self)
	{
		vga::vga_cls();
		self.set_cursor(0,0);
	}

	fn print(&mut self, s: &str)
	{
		for byte in s.bytes() {
			match byte
			{
				b'\n' => self.set_cursor(self.row + 1, 0),
				b'\r' => self.set_cursor(self.row, 0),
				_ => {
						vga::vga_put(byte, self.row, self.col);
						self.move_cursor(1);
				},
			}
		}
	}
}


impl fmt::Write for TermCursor 
{
    fn write_str(&mut self, s: &str) -> fmt::Result 
	{
        self.print(s);
        Ok(())
    }
}


lazy_static! {
	pub static ref TERM: Mutex<TermCursor> = Mutex::new(TermCursor {
		row: 0,
		col: 0,
	});
}

pub fn TERM_print(s: &str) {
	use core::fmt::Write;
	write!(TERM.lock(), "The numbers are {} and {}", 42, 1.0/3.0).unwrap();
	//TERM.lock().print(s);
}

#[macro_export]
macro_rules! print {
        ($($arg:tt)*) => ($crate::drivers::term::_print(format_args!($($arg)*)));
}

#[macro_export]
macro_rules! println {
    () => (print!("\n"));
    ($($arg:tt)*) => (print!("{}\n", format_args!($($arg)*)));
}

#[doc(hidden)]
pub fn _print(args: fmt::Arguments) {
        use core::fmt::Write;
		use x86_64::instructions::interrupts;
        //wrtie_fmt() is from Write trait
		interrupts::without_interrupts(|| {
			TERM.lock().write_fmt(args).unwrap();
		});
}



